src/ImageSharp/Formats/Webp/Lossy/LossyUtils.cs

```diff
@@ -938,26 +942,57 @@ public static void HFilter8i(Span<byte> u, Span<byte> v, int offset, int stride,
         }
 
-        [MethodImpl(InliningOptions.ShortMethod)]
-        public static uint LoadUv(byte u, byte v) =>
-            (uint)(u | (v << 16)); // We process u and v together stashed into 32bit(16bit each).
-
-        [MethodImpl(InliningOptions.ShortMethod)]
-        public static void YuvToBgr(int y, int u, int v, Span<byte> bgr)
+        public static void Mean16x4(Span<byte> input, Span<uint> dc)
         {
-            bgr[0] = (byte)YuvToB(y, u);
-            bgr[1] = (byte)YuvToG(y, u, v);
-            bgr[2] = (byte)YuvToR(y, v);
-        }
-
-        [MethodImpl(InliningOptions.ShortMethod)]
-        public static int YuvToB(int y, int u) => Clip8(MultHi(y, 19077) + MultHi(u, 33050) - 17685);
-
-        [MethodImpl(InliningOptions.ShortMethod)]
-        public static int YuvToG(int y, int u, int v) => Clip8(MultHi(y, 19077) - MultHi(u, 6419) - MultHi(v, 13320) + 8708);
+#if SUPPORTS_RUNTIME_INTRINSICS
+            if (Sse2.IsSupported)
+            {
+                Vector128<byte> a0 = Unsafe.As<byte, Vector128<byte>>(ref MemoryMarshal.GetReference(input));
+                Vector128<byte> a1 = Unsafe.As<byte, Vector128<byte>>(ref MemoryMarshal.GetReference(input.Slice(WebpConstants.Bps, 16)));
+                Vector128<byte> a2 = Unsafe.As<byte, Vector128<byte>>(ref MemoryMarshal.GetReference(input.Slice(WebpConstants.Bps * 2, 16)));
+                Vector128<byte> a3 = Unsafe.As<byte, Vector128<byte>>(ref MemoryMarshal.GetReference(input.Slice(WebpConstants.Bps * 3, 16)));
+                Vector128<short> b0 = Sse2.ShiftRightLogical(a0.AsInt16(), 8); // hi byte
+                Vector128<short> b1 = Sse2.ShiftRightLogical(a1.AsInt16(), 8);
+                Vector128<short> b2 = Sse2.ShiftRightLogical(a2.AsInt16(), 8);
+                Vector128<short> b3 = Sse2.ShiftRightLogical(a3.AsInt16(), 8);
+                Vector128<byte> c0 = Sse2.And(a0, Mean16x4Mask); // lo byte
+                Vector128<byte> c1 = Sse2.And(a1, Mean16x4Mask);
+                Vector128<byte> c2 = Sse2.And(a2, Mean16x4Mask);
+                Vector128<byte> c3 = Sse2.And(a3, Mean16x4Mask);
+                Vector128<int> d0 = Sse2.Add(b0.AsInt32(), c0.AsInt32());
+                Vector128<int> d1 = Sse2.Add(b1.AsInt32(), c1.AsInt32());
+                Vector128<int> d2 = Sse2.Add(b2.AsInt32(), c2.AsInt32());
+                Vector128<int> d3 = Sse2.Add(b3.AsInt32(), c3.AsInt32());
+                Vector128<int> e0 = Sse2.Add(d0, d1);
+                Vector128<int> e1 = Sse2.Add(d2, d3);
+                Vector128<int> f0 = Sse2.Add(e0, e1);
+                Vector128<short> hadd = Ssse3.HorizontalAdd(f0.AsInt16(), f0.AsInt16());
+                Vector64<short> lower = hadd.GetLower();
+
+                dc[3] = (uint)lower.GetElement(3);
+                dc[2] = (uint)lower.GetElement(2);
+                dc[1] = (uint)lower.GetElement(1);
+                dc[0] = (uint)lower.GetElement(0);
```

Comment:
I'm not sure what does `GetElement` compile to but it can be terribly inefficient. `dc[0...3]` is 64 bits if I'm not mistaken:

```suggestion
                Unsafe.As<uint, Vector64<short>>(ref MemoryMarshal.GetReference(dc)) = lower;
```

Reply:
No sorry, it's actually 128 bits. It needs some widening with intrinsics then.

Reply:
I think this GetLower is not needed at all, now that i think again about it.

What about this: 1452ba00

Reply:
@antonfirsov so you think we should try to avoid `GetElement` entirely?

Reply:
I think it!s a good practice to avoid it whenever possible. It looks like it compiles to [VPEXTRW](https://www.felixcloutier.com/x86/pextrw) extracting scalar registers from the SIMD register, but then you do scalar conversion for each value. We can get much better results with shuffling:
[SHARPLAB](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdAEoCuMCUYAplQMID2YADlADYsATgGVhANygBjFigDcRUhWr1GzNgElGwzjzFDJM+YvKVaDJqypaEQ2CmnHiplRfXXGdmA6koqADQAOADYFQkUAZmVgimwydjIAbyIyVOiyBhQIADMWCgAWMgBxFgQAUUFWRgAKADUWKQROIVxsQIAeFAALZoQAPjJxOAzYBAAqMiEAShS05MI0xcmAbQiAXTIAXjJqulGp8SoS8sqWGuwpsKW0oWXsDe3d/cPjipYqhGqLq+uV3AedntGAcjqU3h9qrhLrNrrcSACnsCXmDTjUSNCFmkAL4wsi4ygxTI5PKYQqvVEIAD6Ih4EBgdQaTRabU6PSE/UGwxpdPaQI501x81+K3WW0Bz1BJ3eZ0+31xi1u9zFiIQIPJ0vOGOFt3+yr5apRGs+UJ+sOW8L1EvVEPRprIOPCmNSBIy3mJBTIIiZLAZjWarQ63V6AyGI0YEwFTqS8rS9T9zI6fIGAHcoAATPLbEQoFjYKgAVRgtKkAGsADKcZPVUNxpkB1nBqgALR0UyoAEEUPmbBFsNUtb9s7mqF7mj6hMNUxmB6kHUR8bgYqTPd7qbT6bX/Syg+yQ1z17zRgNI4shb9NwnD4wU+nM56c3nC8Xy5Xq8ML/Wd/1m62O12e32M5LIWWS5H+V4IO+jJbomR59NUQgsNkZAALLvM0ACeKEQEI3QQPwko0EhwhnDICFTFMYpTiwdoOliQA==)

I think we are looking for [_mm_unpacklo_epi16](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm_unpacklo_epi16&ig_expand=732) / `Sse2.UnpackLow` interleaving `lower` with zeros to get equivalent `uint` representation, but not sure how would this act with negative `short`-s. Can `hadd` end up having negative values?

```C#
Sse2.UnpackLow(lower, Vector128<short>.Zero)
```

Reply:
That looks indeed much better. I dont think there can be negative values.

Thanks for your help!

Reply:
using `Sse2.Store` looks even a tiny bit better

Reply:
`dc` is stackallocked, so you can pass it as a pointer easily if you want to, though I'm not sure if it's worth the trouble.
The main benefit would be the removal of one `Slice()` on [the call site](https://github.com/SixLabors/ImageSharp/blob/7d8225b59a633b08b51e74bbb960d4d52b420a84/src/ImageSharp/Formats/Webp/Lossy/Vp8EncIterator.cs#L366), the disadvantage is the inconsistency of the method signature (`input` is a span, `dc` is a pointer).

Reply:
I think its not worth it, lets keep the method signature consistent and keep it as it is.